pkg/token/serviceprincipaltoken.go

```diff
@@ -55,28 +61,46 @@ func newServicePrincipalToken(oAuthConfig adal.OAuthConfig, clientID, clientSecr
 	}, nil
 }
 
+// Token fetches an azcore.AccessToken from the Azure SDK and converts it to an adal.Token for use with kubelogin.
 func (p *servicePrincipalToken) Token() (adal.Token, error) {
-	emptyToken := adal.Token{}
-	callback := func(t adal.Token) error {
-		return nil
-	}
+	return p.TokenOptions(nil)
+}
 
-	var (
-		spt *adal.ServicePrincipalToken
-		err error
-	)
+func (p *servicePrincipalToken) TokenOptions(options *azcore.ClientOptions) (adal.Token, error) {
+	emptyToken := adal.Token{}
+	var spnAccessToken azcore.AccessToken
 
+	// Request a new Azure token provider for service principal
 	if p.clientSecret != "" {
-		spt, err = adal.NewServicePrincipalToken(
-			p.oAuthConfig,
+		clientOptions := &azidentity.ClientSecretCredentialOptions{}
+		if options == nil {
```

Comment:
can be simplified to
```go
		clientOptions := &azidentity.ClientSecretCredentialOptions{}
		if options != nil {
			clientOptions.ClientOptions = *options
		}
```

Reply:
besides, https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore@v1.6.0/policy#ClientOptions is where you would specify [cloud](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore@v1.6.0/cloud).[Configuration](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore@v1.6.0/cloud#Configuration). So you also need to translate that from adal to new sdk

Reply:
Fixed the formatting. Will look into the cloud.Configuration

Reply:
Added cloud.Configuration to provider.go and rebased service principal off of it instead of OAuthConfig